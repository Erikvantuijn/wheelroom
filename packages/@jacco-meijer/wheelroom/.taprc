test-regex: (dist.*\/tests\/.*\.(int|unit?)\.test)\.(mjs|js|cjs|ts)$
