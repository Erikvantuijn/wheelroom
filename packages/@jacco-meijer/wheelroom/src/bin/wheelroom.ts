#!/usr/bin/env node

import * as dotenv from 'dotenv'
import { Argv, CommandBuilder } from 'yargs'
import * as yargs from 'yargs'
import { createContent } from '../commands/create-content/create-content'
import { createFragmentFiles } from '../commands/create-files/fragment/create-fragment-files'
import { createNewModel } from '../commands/create-files/new-model/create-new-model'
import { createModels } from '../commands/create-models/create-models'
import { deleteContent } from '../commands/delete-content/delete-content'
import { list } from '../commands/list/list'
import { getComponentConfigs } from '../config/config'
import { ContentfulApiContext } from '../types/contentful-api-context'

const dotEnvResult = dotenv.config()
if (dotEnvResult.error) {
  throw dotEnvResult.error
}

const cmdList = async (args: { filter: string }) => {
  const componentConfigs = await getComponentConfigs({ filter: args.filter })
  await list(componentConfigs)
}

const cmdCreateModels = async (args: { filter: string }) => {
  const context = {
    componentConfigs: await getComponentConfigs({ filter: args.filter }),
  } as ContentfulApiContext
  await createModels(context)
}

const cmdCreateContent = async (args: { filter: string }) => {
  const context = {
    componentConfigs: await getComponentConfigs({ filter: args.filter }),
  } as ContentfulApiContext
  await createContent(context)
}

const cmdDeleteContent = async (args: { filter: string }) => {
  const context = {
    componentConfigs: await getComponentConfigs({ filter: args.filter }),
  } as ContentfulApiContext
  await deleteContent(context)
}

const cmdCreateFragments = async (args: { filter: string; path: string }) => {
  const componentConfigs = await getComponentConfigs({ filter: args.filter })
  await createFragmentFiles(componentConfigs, args.path)
}

const builderCreateFragments: CommandBuilder = (yargs2: Argv) =>
  yargs2.positional('path', {
    describe: 'where to create the files',
    type: 'string',
  })

const cmdCreateNewModel = async (args: {
  filter: string
  componentPath: string
  configPath: string
}) => {
  await createNewModel(args.componentPath, args.configPath)
}

const builderCreateNewModel: CommandBuilder = (yargs2: Argv) =>
  yargs2
    .positional('component-path', {
      describe: 'where to create the component files',
      type: 'string',
    })
    .positional('config-path', {
      describe: 'where to create the component config files',
      type: 'string',
    })

let params = yargs
  .command({
    command: 'list [options]',
    describe: 'list configured models',
    handler: cmdList,
  })
  .command({
    command: 'create-models [options]',
    describe: 'create or update configured models',
    handler: cmdCreateModels,
  })
  .command({
    command: 'create-content [options]',
    describe: 'create demo content for configured models',
    handler: cmdCreateContent,
  })
  .command({
    command: 'delete-content [options]',
    describe: 'remove all demo content for configured models',
    handler: cmdDeleteContent,
  })
  .command({
    builder: builderCreateFragments,
    command: 'create-fragments <path> [options]',
    describe: 'create files with required graphql fragments',
    handler: cmdCreateFragments as any,
  })
  .command({
    builder: builderCreateNewModel,
    command: 'create-new-model <component-path> <config-path> [options]',
    describe: 'create all files needed for setting up a new model',
    handler: cmdCreateNewModel as any,
  })
  .command({
    command: '*',
    handler: () => {
      console.log(`Use --help flag for options`)
    },
  })
  .alias('f', 'filter')
  .describe('f', 'Filter by componentType').argv

// Prevent ts warning
params = params
