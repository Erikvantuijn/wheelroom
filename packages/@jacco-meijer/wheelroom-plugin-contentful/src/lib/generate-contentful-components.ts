import {
  Field as WheelroomField,
  WheelroomComponent,
  WheelroomComponents,
} from '@jacco-meijer/wheelroom'
import { getCases } from '@jacco-meijer/wheelroom'
import {
  ContentfulComponent,
  ContentfulComponents,
  Field as ContentfulField,
  Fields as ContentfulFields,
} from '../types/contentful-components'
import { ContentfulFieldDefinitions } from '../types/contentful-field-definitions'

export const generateContentfulComponents = (
  wheelroomComponents: WheelroomComponents,
  fieldDefinitions: ContentfulFieldDefinitions
): ContentfulComponents => {
  console.log('merging', wheelroomComponents)
  console.log('and', fieldDefinitions)
  const components: ContentfulComponents = {}
  // Loop through all configured wheelroom components
  Object.entries(wheelroomComponents).forEach(
    ([componentName, wheelroomComponent]: [string, WheelroomComponent]) => {
      const componentNameCases = getCases(componentName)
      const fields: ContentfulFields = {}
      // For each component, loop through all configured wheelroom fields
      Object.entries(wheelroomComponent.fields).forEach(
        ([fieldName, fieldValue]: [string, WheelroomField]) => {
          if (fieldValue.system) {
            // Skip system fields
            return
          }
          // Create a working copy based on the Contentful field definition
          const workingField = Object.assign(
            {},
            fieldDefinitions.fieldTypes[fieldValue.type!]
          )
          // TODO: Parse all fields
          // TODO: Deep copy into working field

          // Loop through all attributes of a field and map onto contentful working field
          Object.entries(fieldValue).forEach(
            ([attributeName, attributeValue]: [string, string]) => {
              if (attributeName in fieldDefinitions.systemAttributes) {
                // Skip system attributes like 'system' and 'type'
                return
              }
              if (attributeName in fieldDefinitions.attributeMappings) {
                // Map attributes that have a mapping into the Contentful field
                console.log(
                  'path:',
                  fieldDefinitions.attributeMappings[attributeName]
                )
                return
              }
              // All other attributes go into the root of the Contentful field
              workingField[attributeName] = attributeValue
            }
          )
          // Copy all workFields into a new ContentfulField
          // ...
          let field: ContentfulField
          fields[fieldName] = field
        }
      )
      const component: ContentfulComponent = {
        description: componentNameCases.sentenceCase,
        displayField: 'title',
        fields,
        modelVersion: wheelroomComponent.modelVersion,
        type: componentNameCases.camelCase,
      }
      components[componentName] = component
    }
  )
  return components
}
