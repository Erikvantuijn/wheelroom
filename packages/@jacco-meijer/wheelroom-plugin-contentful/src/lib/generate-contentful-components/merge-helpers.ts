import { FieldType } from '@jacco-meijer/wheelroom'

/**
 * - %firstItem%
 * - %firstComponent%
 * - %demoAsset%
 */
export const initialContentParser = (
  unparsed: string,
  field: FieldType
): string => {
  // If objects or arrays are missing in the field, the variable has been used
  // in a field where the value cannot be found. To prevent errors, supply empty
  // values by creating a safe field that has all values.
  const safeField = Object.assign(
    {},
    { items: ['not-found'], components: ['not-found'] },
    field
  )
  switch (unparsed) {
    case '%demoAsset%':
      return 'demoAsset'
    case '%firstItem%':
      return safeField.items[0]
    case '%firstComponent%':
      return safeField.components[0]
  }
  return unparsed
}

// Helper to add object or array if needed
export const createIfMissing = (
  base: any,
  key: string,
  what: 'object' | 'array'
) => {
  if (key in base) {
    return base
  } else {
    base[key] = what === 'object' ? {} : []
    return base
  }
}
