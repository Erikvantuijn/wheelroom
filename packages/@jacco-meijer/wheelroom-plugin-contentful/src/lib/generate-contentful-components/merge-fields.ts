/**
 * mergeFields
 *
 * Take a wheelroom field and convert it to a parsed Contentful field
 *
 * Process Wheelroom fields:
 *
 * - helpText?: string -> settings.helpText
 * - initialContent?:  string | number | boolean | string[] -> createContentData
 * - localized?: boolean -> specs.localized
 * - name?: string -> specs.name
 * - required?: boolean -> specs.required
 * - system?: boolean -> SKIP
 * - type?: FieldType -> REPLACE WITH DEFINITION
 *
 *
 * Process validations:
 *
 * - components: string[] -> specs.validations.0.linkContentType
 * -                      -> specs.items.validations.0.linkContentType
 * - items?: string[] -> specs.validations.0.in
 * - maxLength?: number -> specs.validations.0.size.max
 * - unique?: boolean -> specs.validations.0.unique
 *
 */

import { FieldType, parser, WheelroomComponents } from '@jacco-meijer/wheelroom'
import { ContentfulField, widgetID } from '../../types/contentful-fields'
import { createContentDataParser, createIfMissing } from './merge-helpers'

interface MergeFields {
  cfFieldDefaults: ContentfulField
  cfFieldDefinition: ContentfulField
  componentName: string
  fieldName: string
  wrComponents: WheelroomComponents
  wrField: FieldType
}

export const mergeFields = (context: MergeFields): ContentfulField => {
  // Hard coded system defaults as a last resort
  const lastResort = {
    helpText: '',
    localized: false,
    required: false,
    type: 'Symbol',
    widgetId: 'singleLine' as widgetID,
  }

  // Add optional settings object if not present
  if (!('settings' in context.cfFieldDefinition)) {
    context.cfFieldDefinition.settings = {}
  }
  if (!('settings' in context.cfFieldDefaults)) {
    context.cfFieldDefaults.settings = {}
  }
  // Parser shortcut with commandName and context.fieldName
  const preLoadedParser = (unparsed: string) =>
    parser(unparsed, {
      componentName: context.componentName,
      fieldName: context.fieldName,
    })

  const workingField: ContentfulField = {
    settings: {
      helpText: preLoadedParser(
        context.wrField.helpText ||
          context.cfFieldDefinition.settings!.helpText ||
          context.cfFieldDefaults.settings!.helpText ||
          lastResort.helpText
      ),
    },
    specs: {
      localized:
        context.wrField.localized ||
        context.cfFieldDefinition.specs.localized ||
        context.cfFieldDefaults.specs.localized ||
        lastResort.localized,
      name: preLoadedParser(
        context.wrField.name ||
          context.cfFieldDefinition.specs.name ||
          context.cfFieldDefaults.specs.name
      ),
      required:
        context.wrField.required ||
        context.cfFieldDefinition.specs.required ||
        context.cfFieldDefaults.specs.required ||
        lastResort.required,
      type:
        context.cfFieldDefinition.specs.type ||
        context.cfFieldDefaults.specs.type ||
        lastResort.type,
    },
    widgetId:
      context.cfFieldDefinition.widgetId ||
      context.cfFieldDefaults.widgetId ||
      lastResort.widgetId,
  }

  // Handle initial content, a if-then-tree because of the complex union type
  let createContentData
  if (context.wrField.initialContent) {
    createContentData = context.wrField.initialContent
    // }
  } else if (context.cfFieldDefinition.createContentData) {
    // No initial content configured, lookup field definition
    createContentData = context.cfFieldDefinition.createContentData
  } else if (context.cfFieldDefaults.createContentData) {
    // No value in field definition, check default values
    createContentData = context.cfFieldDefaults.createContentData
  }
  if (typeof createContentData === 'string') {
    createContentData = createContentDataParser(
      createContentData,
      context.wrField
    )
    createContentData = preLoadedParser(createContentData as string)
  }
  Object.assign(workingField, { createContentData })

  // Handle linkType
  const linkType =
    context.cfFieldDefinition.specs.linkType ||
    context.cfFieldDefaults.specs.linkType
  if (linkType) {
    workingField.specs.linkType = linkType
  }

  // Handle items
  const items =
    context.cfFieldDefinition.specs.items || context.cfFieldDefaults.specs.items
  if (items) {
    workingField.specs.items = Object.assign({}, items)
  }

  // Handle validations present in context.cfFieldDefinition or context.cfFieldDefaults components
  const validations =
    context.cfFieldDefinition.specs.validations ||
    context.cfFieldDefaults.specs.validations
  if (validations) {
    createIfMissing(workingField.specs, 'validations', 'array')
    validations.forEach((validation: any) => {
      workingField.specs.validations!.push(Object.assign({}, validation))
    })
  }
  // Process other validations
  // singleComponent: specs.validations.0.linkContentType
  if (context.wrField.type === 'singleComponent') {
    createIfMissing(workingField.specs, 'validations', 'array')
    workingField.specs.validations!.push({
      linkContentType: context.wrField.allowedComponents,
    })
  }

  // multipleComponents: specs.items.validations.0.linkContentType
  if (context.wrField.type === 'multipleComponents') {
    createIfMissing(workingField.specs, 'items', 'object')
    createIfMissing(workingField.specs.items, 'validations', 'array')
    workingField.specs.items!.validations!.push({
      linkContentType: context.wrField.allowedComponents,
    })
  }

  if (context.wrField.type === 'dropdown') {
    // specs.validations.0.in
    createIfMissing(workingField.specs, 'validations', 'array')
    workingField.specs.validations!.push({
      in: context.wrField.items,
    })
  }
  if (context.wrField.type === 'shortText' && context.wrField.maxLength) {
    // specs.validations.0.size.max
    createIfMissing(workingField.specs, 'validations', 'array')
    workingField.specs.validations!.push({
      size: {
        max: context.wrField.maxLength,
        min: 0,
      },
    })
  }
  if (context.wrField.unique) {
    // specs.validations.0.unique
    createIfMissing(workingField.specs, 'validations', 'array')
    workingField.specs.validations!.push({
      unique: context.wrField.unique,
    })
  }

  return workingField
}
