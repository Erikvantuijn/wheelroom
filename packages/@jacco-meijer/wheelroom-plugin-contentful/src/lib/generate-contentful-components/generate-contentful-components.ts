import {
  FieldType,
  WheelroomComponent,
  WheelroomComponents,
} from '@jacco-meijer/wheelroom'
import { getCases } from '@jacco-meijer/wheelroom'
import {
  ContentfulComponent,
  ContentfulComponents,
} from '../../types/contentful-components'
import { ContentfulFieldDefinitions } from '../../types/contentful-field-definitions'
import { ContentfulFields } from '../../types/contentful-fields'
import { mergeFields } from './merge-fields'

export const generateContentfulComponents = (
  wheelroomComponents: WheelroomComponents,
  fieldDefinitions: ContentfulFieldDefinitions
): ContentfulComponents => {
  const components: ContentfulComponents = {}
  // Loop through all configured wheelroom components
  Object.entries(wheelroomComponents).forEach(
    ([componentName, wheelroomComponent]: [string, WheelroomComponent]) => {
      const componentNameCases = getCases(componentName)
      const fields: ContentfulFields = {}
      // For each component, loop through all configured wheelroom fields
      Object.entries(wheelroomComponent.fields).forEach(
        ([fieldName, fieldValue]: [string, FieldType]) => {
          if (fieldValue.system) {
            // Skip system fields
            return
          }
          const contentfulFieldDefinition =
            fieldDefinitions.fieldTypes[fieldValue.type!]
          fields[fieldName] = mergeFields({
            cfFieldDefinition: contentfulFieldDefinition,
            componentName,
            fieldName,
            wrComponents: wheelroomComponents,
            wrField: fieldValue,
          })
        }
      )
      const component: ContentfulComponent = {
        description: componentNameCases.sentenceCase,
        displayField: 'title',
        fields,
        modelVersion: wheelroomComponent.modelVersion,
        type: componentNameCases.camelCase,
      }
      components[componentName] = component
    }
  )
  return components
}
