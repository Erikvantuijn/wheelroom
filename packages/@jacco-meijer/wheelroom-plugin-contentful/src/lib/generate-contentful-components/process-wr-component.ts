import {
  FieldType,
  WheelroomComponent,
  WheelroomComponents,
} from '@jacco-meijer/wheelroom'
import { getCases } from '@jacco-meijer/wheelroom'
import { ContentfulComponent } from '../../types/contentful-components'
import { ContentfulFieldDefinitions } from '../../types/contentful-field-definitions'
import { ContentfulFields } from '../../types/contentful-fields'
import { mergeFields } from './merge-fields'

interface ProcessWrComponent {
  componentName: string
  fieldDefinitions: ContentfulFieldDefinitions
  wheelroomComponent: WheelroomComponent
  wheelroomComponents: WheelroomComponents
}

export const processWrComponent = (context: ProcessWrComponent) => {
  const componentNameCases = getCases(context.componentName)
  const fields: ContentfulFields = {}
  // For each component, loop through all configured wheelroom fields
  Object.entries(context.wheelroomComponent.fields).forEach(
    ([fieldName, fieldValue]: [string, FieldType]) => {
      if (fieldValue.system) {
        // Skip system fields
        return
      }
      let contentfulFieldDefinition
      // If field.typePostfix is present, use it
      if (
        'typePostfix' in fieldValue &&
        fieldValue.type + fieldValue.typePostfix in
          context.fieldDefinitions.fieldTypes
      ) {
        contentfulFieldDefinition =
          context.fieldDefinitions.fieldTypes[
            fieldValue.type + fieldValue.typePostfix
          ]
      } else {
        // Field.typePostfix not present, use field.type
        contentfulFieldDefinition =
          context.fieldDefinitions.fieldTypes[fieldValue.type!]
      }
      fields[fieldName] = mergeFields({
        cfFieldDefinition: contentfulFieldDefinition,
        componentName: context.componentName,
        fieldName,
        wrComponents: context.wheelroomComponents,
        wrField: fieldValue,
      })
    }
  )
  const component: ContentfulComponent = {
    description: componentNameCases.sentenceCase,
    displayField: 'title',
    fields,
    modelVersion: context.wheelroomComponent.modelVersion,
    type: componentNameCases.camelCase,
  }
  return component
}
