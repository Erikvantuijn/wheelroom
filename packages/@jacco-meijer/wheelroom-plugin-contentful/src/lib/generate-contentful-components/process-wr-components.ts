import {
  FieldType,
  WheelroomComponent,
  WheelroomComponents,
} from '@jacco-meijer/wheelroom'
import { getCases } from '@jacco-meijer/wheelroom'
import {
  ContentfulComponent,
  ContentfulComponents,
} from '../../types/contentful-components'
import { ContentfulFieldDefinitions } from '../../types/contentful-field-definitions'
import { ContentfulFields } from '../../types/contentful-fields'
import { ContentSet } from './generate-contentful-components'
import { mergeFields } from './merge-fields'

interface ProcessWrComponents {
  componentName: string
  components: ContentfulComponents
  contentSet: ContentSet | undefined
  fieldDefinitions: ContentfulFieldDefinitions
  wheelroomComponent: WheelroomComponent
  wheelroomComponents: WheelroomComponents
}

export const processWrComponents = (context: ProcessWrComponents) => {
  const componentNameCases = getCases(context.componentName)
  const fields: ContentfulFields = {}
  // For each component, loop through all configured wheelroom fields
  Object.entries(context.wheelroomComponent.fields).forEach(
    ([fieldName, fieldValue]: [string, FieldType]) => {
      if (fieldValue.system) {
        // Skip system fields
        return
      }
      let contentfulFieldDefinition
      // If field.typePostfix is present, use it
      if (
        'typePostfix' in fieldValue &&
        fieldValue.type + fieldValue.typePostfix in
          context.fieldDefinitions.fieldTypes
      ) {
        contentfulFieldDefinition =
          context.fieldDefinitions.fieldTypes[
            fieldValue.type + fieldValue.typePostfix
          ]
      } else {
        // Field.typePostfix not present, use field.type
        contentfulFieldDefinition =
          context.fieldDefinitions.fieldTypes[fieldValue.type!]
      }
      fields[fieldName] = mergeFields({
        cfFieldDefinition: contentfulFieldDefinition,
        componentName: context.componentName,
        fieldName,
        wrComponents: context.wheelroomComponents,
        wrField: fieldValue,
      })
    }
  )
  const component: ContentfulComponent = {
    description: componentNameCases.sentenceCase,
    displayField: 'title',
    fields,
    modelVersion: context.wheelroomComponent.modelVersion,
    type: componentNameCases.camelCase,
  }
  context.components[context.componentName] = component
}
