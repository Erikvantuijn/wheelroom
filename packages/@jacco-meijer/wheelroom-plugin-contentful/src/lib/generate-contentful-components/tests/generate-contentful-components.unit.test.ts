import { getFilteredComponents } from '@jacco-meijer/wheelroom'
import { argvCommand } from '../../../fixtures/argv-command'
import { generateContentfulComponents } from '../generate-contentful-components'
import { output } from './output'

describe('Generate contentful components should', () => {
  test('match the sample sets', () => {
    const pluginOptions =
      argvCommand.options['@jacco-meijer/wheelroom-plugin-contentful']
    const wheelroomComponents = getFilteredComponents(argvCommand)

    const contentfulComponents = generateContentfulComponents(
      wheelroomComponents,
      pluginOptions.fieldDefinitions
    )
    // console.log('OUTPUT', JSON.stringify(contentfulComponents, null, 2))
    expect(contentfulComponents).toStrictEqual(output)
  })
})
