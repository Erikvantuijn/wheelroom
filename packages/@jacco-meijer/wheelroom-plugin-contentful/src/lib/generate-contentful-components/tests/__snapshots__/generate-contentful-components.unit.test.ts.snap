// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate contentful components should match the sample sets 1`] = `
Object {
  "allFieldTypes": Object {
    "description": "All field types",
    "displayField": "title",
    "fields": Object {
      "dateField": Object {
        "createContentData": "2019-06-03T00:00+01:00",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Date field",
          "required": false,
          "type": "Date",
        },
        "widgetId": "datePicker",
      },
      "dropdownField": Object {
        "createContentData": "item A",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Dropdown field",
          "required": false,
          "type": "Symbol",
          "validations": Array [
            Object {
              "in": Array [
                "item A",
                "item B",
                "item C",
              ],
            },
          ],
        },
        "widgetId": "dropdown",
      },
      "imageField": Object {
        "createContentData": "demoAsset",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "linkType": "Asset",
          "localized": false,
          "name": "Image field",
          "required": false,
          "type": "Link",
          "validations": Array [
            Object {
              "linkMimetypeGroup": Array [
                "image",
              ],
            },
          ],
        },
        "widgetId": "assetLinkEditor",
      },
      "longTextField": Object {
        "createContentData": "This is a long text field",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Long text field",
          "required": false,
          "type": "Text",
        },
        "widgetId": "multipleLine",
      },
      "multipleComponentsField": Object {
        "createContentData": Array [
          "firstSection",
          "secondSection",
        ],
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "items": Object {
            "linkType": "Entry",
            "type": "Link",
            "validations": Array [
              Object {
                "linkContentType": Array [
                  "page",
                  "firstSection",
                  "secondSection",
                ],
              },
            ],
          },
          "localized": false,
          "name": "Multiple components field",
          "required": false,
          "type": "Array",
        },
        "widgetId": "entryLinksEditor",
      },
      "numberField": Object {
        "createContentData": 5,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Number field",
          "required": false,
          "type": "Integer",
        },
        "widgetId": "numberEditor",
      },
      "richTextField": Object {
        "createContentData": "This is a rich text field",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Rich text field",
          "required": false,
          "type": "RichText",
        },
        "widgetId": "richTextEditor",
      },
      "shortTextField": Object {
        "createContentData": "This is a short text field",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Short text field",
          "required": false,
          "type": "Symbol",
          "validations": Array [
            Object {
              "size": Object {
                "max": 155,
                "min": 0,
              },
            },
          ],
        },
        "widgetId": "singleLine",
      },
      "singleComponentField": Object {
        "createContentData": "pageModel",
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "linkType": "Entry",
          "localized": false,
          "name": "Single component field",
          "required": false,
          "type": "Link",
          "validations": Array [
            Object {
              "linkContentType": Array [
                "pageModel",
              ],
            },
          ],
        },
        "widgetId": "entryLinkEditor",
      },
      "tagsField": Object {
        "createContentData": Array [
          "tag A",
          "tag B",
        ],
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "items": Object {
            "type": "Symbol",
          },
          "localized": false,
          "name": "Tags field",
          "required": false,
          "type": "Array",
        },
        "widgetId": "tagEditor",
      },
      "title": Object {
        "createContentData": "All field types generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "allFieldTypes",
  },
  "firstSection": Object {
    "description": "First section",
    "displayField": "title",
    "fields": Object {
      "defaultField": Object {
        "createContentData": undefined,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Default field",
          "required": false,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
      "title": Object {
        "createContentData": "First section generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "firstSection",
  },
  "globalModel": Object {
    "description": "Global model",
    "displayField": "title",
    "fields": Object {
      "heading": Object {
        "createContentData": undefined,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Heading",
          "required": false,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
      "title": Object {
        "createContentData": "Global model generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "globalModel",
  },
  "noFragment": Object {
    "description": "No fragment",
    "displayField": "title",
    "fields": Object {
      "heading": Object {
        "createContentData": undefined,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Heading",
          "required": false,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
      "title": Object {
        "createContentData": "No fragment generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "noFragment",
  },
  "pageModel": Object {
    "description": "Page model",
    "displayField": "title",
    "fields": Object {
      "heading": Object {
        "createContentData": undefined,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Heading",
          "required": false,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
      "title": Object {
        "createContentData": "Page model generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "pageModel",
  },
  "secondSection": Object {
    "description": "Second section",
    "displayField": "title",
    "fields": Object {
      "heading": Object {
        "createContentData": undefined,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Heading",
          "required": false,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
      "title": Object {
        "createContentData": "Second section generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "secondSection",
  },
  "subPageModel": Object {
    "description": "Sub page model",
    "displayField": "title",
    "fields": Object {
      "heading": Object {
        "createContentData": undefined,
        "settings": Object {
          "helpText": "",
        },
        "specs": Object {
          "localized": false,
          "name": "Heading",
          "required": false,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
      "title": Object {
        "createContentData": "Sub page model generated from boilerplate",
        "settings": Object {
          "helpText": "Never displayed, only used for listing within Contentful",
        },
        "specs": Object {
          "localized": false,
          "name": "Title",
          "required": true,
          "type": "Symbol",
        },
        "widgetId": "singleLine",
      },
    },
    "modelVersion": "1.0.0",
    "type": "subPageModel",
  },
}
`;
